bindings/wasm/src/did/wasm_core_document.rs

```diff
@@ -513,11 +583,45 @@ struct ICoreDocumentHelper {
   properties: Object,
 }
 
-impl_wasm_json!(WasmCoreDocument, CoreDocument);
-impl_wasm_clone!(WasmCoreDocument, CoreDocument);
+#[wasm_bindgen(js_class = CoreDocument)]
+impl WasmCoreDocument {
+  /// Serializes this to a JSON object.
+  #[wasm_bindgen(js_name = toJSON)]
+  pub fn to_json(&self) -> crate::error::Result<JsValue> {
+    use crate::error::WasmResult;
+    JsValue::from_serde(&self.0.blocking_read().as_ref()).wasm_result()
+  }
+
+  /// Deserializes an instance from a JSON object.
+  #[wasm_bindgen(js_name = fromJSON)]
+  pub fn from_json(json: &JsValue) -> crate::error::Result<WasmCoreDocument> {
+    use crate::error::WasmResult;
+    json
+      .into_serde()
+      .map(|value| Self(Rc::new(CoreDocumentLock::new(value))))
+      .wasm_result()
+  }
+}
 
 impl From<CoreDocument> for WasmCoreDocument {
   fn from(doc: CoreDocument) -> Self {
-    Self(doc)
+    Self(Rc::new(CoreDocumentLock::new(doc)))
   }
 }
+
+#[wasm_bindgen]
+extern "C" {
+  #[wasm_bindgen(typescript_type = "CoreDocument | IAsCoreDocument")]
+  pub type IAsCoreDocument;
+
+  #[wasm_bindgen(typescript_type = "Array<CoreDocument | IAsCoreDocument>")]
+  pub type ArrayIAsCoreDocument;
+}
+
+#[wasm_bindgen(typescript_custom_section)]
+pub const TS_AS_REF_CORE_Document: &'static str = r#"
+interface IAsCoreDocument {
```

Comment:
I think the more conventional way to call this would be `IToCoreDocument` in the JS/TS world, as everything is `to` there. This feels a bit like leaking Rust terminology to TS. But not strongly opposed to it and if this results in a ton of renaming then let's leave it as-is.

Reply:
@eike-hass Would like your input on this one.

Reply:
The `to` notation is mostly introduced by bindgen, but I think we have precedence for `to` then `as` so let'S go with what is more common in our codebase.